File: impl/ts/src/tool/grep.ts
// grep — regex content search over the jailed project tree.
// See docs/harness-design.md §5.2. Pure-JS (no dependency on rg being present
// in the sandbox); walks files via walk.ts and matches each line.

import type { ToolDef, ToolContext, ToolResult } from "./types.js";
import { reqString, optString } from "./validate.js";
import { walkFiles } from "./walk.js";
import { MAX_LINE_LENGTH } from "./truncate.js";

interface Args {
  pattern: string;
  path?: string;
  include?: string;
}

const MATCH_LIMIT = 100;

// Translate a simple include pattern (e.g. "*.tsx", "*.{ts,tsx}") to a RegExp.
function includeMatcher(include: string | undefined): (p: string) => boolean {
  if (!include) return () => true;
  // Expand {a,b} alternations, then * → [^/]*, ? → [^/].
  const expanded = include.replace(/\{([^}]*)\}/g, (_m, inner: string) => `(${inner.split(",").join("|")})`);
  const re = new RegExp(
    "^" + expanded.replace(/[.+^${}()|[\]\\]/g, (c) => (c === "{" || c === "}" || c === "(" || c === ")" || c === "|" ? c : "\\" + c)).replace(/\*/g, "[^/]*").replace(/\?/g, "[^/]") + "$",
  );
  return (p: string) => re.test(p.split("/").pop() ?? p);
}

export const GrepTool: ToolDef<Args> = {
  id: "grep",
  description:
    "Search file contents by regular expression across the project. Returns file paths with line numbers and the matching lines. Use include to filter files (e.g. \"*.tsx\").",
  parameters: {
    pattern: { type: "string", description: "Regex to search for in file contents", required: true },
    path: { type: "string", description: "Directory to search under (default: project root)" },
    include: { type: "string", description: 'File filter, e.g. "*.tsx" or "*.{ts,tsx}"' },
  },
  validate(args) {
    return { pattern: reqString(args, "pattern"), path: optString(args, "path"), include: optString(args, "include") };
  },
  async execute(args, ctx: ToolContext): Promise<ToolResult> {
    let re: RegExp;
    try {
      re = new RegExp(args.pattern);
    } catch (e) {
      return {
        title: args.pattern,
        output: `Invalid regex: ${String((e as Error).message ?? e)}`,
        metadata: { error: "bad_regex" },
      };
    }
    const matchFile = includeMatcher(args.include);
    const files = walkFiles(ctx, args.path ?? ".").filter(matchFile);

    const perFile: Array<{ path: string; lines: Array<{ n: number; text: string }> }> = [];
    let total = 0;
    let truncated = false;
    for (const f of files) {
      let content: string;
      try {
        content = ctx.store.readFile(ctx.projectId, f);
      } catch {
        continue;
      }
      const lines = content.split("\n");
      const hits: Array<{ n: number; text: string }> = [];
      for (let i = 0; i < lines.length; i++) {
        if (re.test(lines[i])) {
          total++;
          if (total > MATCH_LIMIT) {
            truncated = true;
            break;
          }
          const text = lines[i].length > MAX_LINE_LENGTH ? lines[i].slice(0, MAX_LINE_LENGTH) + " …" : lines[i];
          hits.push({ n: i + 1, text });
        }
      }
      if (hits.length > 0) perFile.push({ path: f, lines: hits });
      if (truncated) break;
    }

    if (total === 0) {
      return { title: args.pattern, output: "No files found", metadata: { matches: 0 } };
    }
    const header = `Found ${total} matches${truncated ? ` (showing first ${MATCH_LIMIT})` : ""}`;
    const body = perFile
      .map((pf) => `\n${pf.path}:\n` + pf.lines.map((l) => `  Line ${l.n}: ${l.text}`).join("\n"))
      .join("\n");
    return {
      title: args.pattern,
      output: header + "\n" + body,
      metadata: { matches: total, truncated },
    };
  },
};

File: impl/go/internal/tool/grep.go
// grep — regex content search over the jailed project tree (Go parity of grep.ts).
package tool

import (
	"fmt"
	"regexp"
	"strings"
)

const grepMatchLimit = 100

// includeMatcher translates a simple include pattern (e.g. "*.tsx",
// "*.{ts,tsx}") to a basename matcher. Empty include matches everything.
func includeMatcher(include string) func(string) bool {
	if include == "" {
		return func(string) bool { return true }
	}
	// Expand {a,b} -> (a|b), then escape regex specials except the alternation,
	// then * -> [^/]*, ? -> [^/].
	expanded := regexp.MustCompile(`\{([^}]*)\}`).ReplaceAllStringFunc(include, func(m string) string {
		inner := strings.TrimSuffix(strings.TrimPrefix(m, "{"), "}")
		return "(" + strings.ReplaceAll(inner, ",", "|") + ")"
	})
	var b strings.Builder
	b.WriteString("^")
	for _, r := range expanded {
		switch r {
		case '*':
			b.WriteString("[^/]*")
		case '?':
			b.WriteString("[^/]")
		case '(', ')', '|':
			b.WriteRune(r)
		case '.', '+', '^', '$', '{', '}', '[', ']', '\\':
			b.WriteByte('\\')
			b.WriteRune(r)
		default:
			b.WriteRune(r)
		}
	}
	b.WriteString("$")
	re, err := regexp.Compile(b.String())
	if err != nil {
		return func(string) bool { return true }
	}
	return func(p string) bool {
		base := p
		if i := strings.LastIndex(p, "/"); i >= 0 {
			base = p[i+1:]
		}
		return re.MatchString(base)
	}
}

// GrepTool searches file contents by regular expression.
var GrepTool = Def{
	ID:          "grep",
	Description: `Search file contents by regular expression across the project. Returns file paths with line numbers and the matching lines. Use include to filter files (e.g. "*.tsx").`,
	Parameters: []Param{
		{Name: "pattern", Type: "string", Description: "Regex to search for in file contents", Required: true},
		{Name: "path", Type: "string", Description: "Directory to search under (default: project root)"},
		{Name: "include", Type: "string", Description: `File filter, e.g. "*.tsx" or "*.{ts,tsx}"`},
	},
	Validate: func(args map[string]any) (map[string]any, error) {
		pat, err := reqString(args, "pattern")
		if err != nil {
			return nil, err
		}
		p, err := optString(args, "path")
		if err != nil {
			return nil, err
		}
		inc, err := optString(args, "include")
		if err != nil {
			return nil, err
		}
		return map[string]any{"pattern": pat, "path": p, "include": inc}, nil
	},
	Execute: func(args map[string]any, ctx *Context) (Result, error) {
		pattern := args["pattern"].(string)
		root := args["path"].(string)
		if root == "" {
			root = "."
		}
		re, err := regexp.Compile(pattern)
		if err != nil {
			return Result{Title: pattern, Output: "Invalid regex: " + err.Error(), Metadata: map[string]any{"error": "bad_regex"}}, nil
		}
		matchFile := includeMatcher(args["include"].(string))

		type fileHits struct {
			path  string
			lines []string
		}
		var perFile []fileHits
		total := 0
		truncated := false
		for _, f := range walkFiles(ctx, root, 5000) {
			if !matchFile(f) {
				continue
			}
			content, err := ctx.Store.ReadFile(ctx.ProjectID, f)
			if err != nil {
				continue
			}
			var hits []string
			for i, line := range strings.Split(content, "\n") {
				if re.MatchString(line) {
					total++
					if total > grepMatchLimit {
						truncated = true
						break
					}
					if len(line) > maxLineLength {
						line = line[:maxLineLength] + " …"
					}
					hits = append(hits, fmt.Sprintf("  Line %d: %s", i+1, line))
				}
			}
			if len(hits) > 0 {
				perFile = append(perFile, fileHits{f, hits})
			}
			if truncated {
				break
			}
		}
		if total == 0 {
			return Result{Title: pattern, Output: "No files found", Metadata: map[string]any{"matches": 0}}, nil
		}
		header := fmt.Sprintf("Found %d matches", total)
		if truncated {
			header += fmt.Sprintf(" (showing first %d)", grepMatchLimit)
		}
		var b strings.Builder
		b.WriteString(header)
		for _, pf := range perFile {
			b.WriteString("\n\n" + pf.path + ":\n" + strings.Join(pf.lines, "\n"))
		}
		return Result{Title: pattern, Output: b.String(), Metadata: map[string]any{"matches": total, "truncated": truncated}}, nil
	},
}
